src/InterchainCreate2FactoryRouter.sol

```diff
@@ -0,0 +1,280 @@
+ */
+contract InterchainCreate2FactoryRouter is Router {
+    // ============ Libraries ============
+
+    using TypeCasts for address;
+    using TypeCasts for bytes32;
+
+    // ============ Constants ============
+
+    // ============ Public Storage ============
+
+    // ============ Upgrade Gap ============
+
+    uint256[47] private __GAP;
+
+    // ============ Events ============
+
+    event RemoteDeployDispatched(uint32 indexed destination, address indexed owner, uint32 _destination, bytes32 ism);
+
+    event Deployed(bytes32 indexed bytecodeHash, bytes32 indexed salt, address indexed deployedAddress);
+
+    // ============ Constructor ============
+
+    constructor(address _mailbox) Router(_mailbox) { }
+
+    // ============ Initializers ============
+
+    /**
+     * @notice Initializes the contract with HyperlaneConnectionClient contracts
+     * @param _domains The domains of the remote Application Routers
+     * @param _customHook used by the Router to set the hook to override with
+     * @param _interchainSecurityModule The address of the local ISM contract
+     * @param _owner The address with owner privileges
+     */
+    function initialize(
+        uint32[] calldata _domains,
+        address _customHook,
+        address _interchainSecurityModule,
+        address _owner
+    )
+        external
+        initializer
+    {
+        _MailboxClient_initialize(_customHook, _interchainSecurityModule, _owner);
+
+        uint256 length = _domains.length;
+        for (uint256 i = 0; i < length; i += 1) {
+            _enrollRemoteRouter(_domains[i], TypeCasts.addressToBytes32(address(this)));
+        }
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Register the address of a Router contract with the same address as this for the same Application on a
+     * remote chain
+     * @param _domain The domain of the remote Application Router
+     */
+    function enrollRemoteDomain(uint32 _domain) external virtual onlyOwner {
+        _enrollRemoteRouter(_domain, TypeCasts.addressToBytes32(address(this)));
+    }
+
+    /**
+     * @notice Batch version of `enrollRemoteDomain
+     * @param _domains The domains of the remote Application Routers
+     */
+    function enrollRemoteDomains(uint32[] calldata _domains) external virtual onlyOwner {
+        uint256 length = _domains.length;
+        for (uint256 i = 0; i < length; i += 1) {
+            _enrollRemoteRouter(_domains[i], TypeCasts.addressToBytes32(address(this)));
+        }
+    }
+
+    /**
+     * @notice Deploys a contract on the `_destination` chain
+     * @param _destination The remote domain
+     * @param _ism The address of the remote ISM, zero address for using the remote mailbox default ISM
+     * @param _salt The salt used for deploying the contract with CREATE2
+     * @param _bytecode The bytecode of the contract to deploy
+     */
+    function deployContract(
```

Comment:
One thing we probably didn't specify is that ideally this router includes remote router overrides like `callRemoteWithOverrides` of InterchainAccountRouter, to allow callers to reuse the existing router address on an existing chain, but not require the owner of the router to populate the routers mapping (since folks can deploy hyperlane themselves).

Reply:
> since folks can deploy hyperlane themselves
So that's what the `Overrides` is for! 
I'll add it